### PR TITLE
[serve] appropriately handle actor died and actor unavailable in all cases

### PR DESCRIPTION
**Pull request by zcin**
## Why are these changes needed?

Sometimes the replicas that a router holds a handle to died, and sometimes they are temporarily unreachable because of network issues. Previously we handled `ActorDiedError` and `ActorUnavailableError` when the router *actively probed* a replica.

Breakdown of scenarios for a replica:
(1) cache is turned off
(2) cache is turned on and replica's cached entry hasn't expired
(3) cache is turned on and replica's cached entry has expired

In case (1) and (3), we actively probe the replica, so the errors are dealt with appropriately. However in case (2), we do not handle the errors so we could repeatedly try to send requests to a dead replica or an unavailable one, and this can persist for up to 10s (or however long the cache timeout is).

This PR adds appropriate handling of `ActorDiedError` and `ActorUnavailableError` upon receiving the system message after sending a request.
- For `ActorDiedError`, the replica will be removed from the router's replica set.
- For `ActorUnavailableError`, the replica's cache entry will be invalidated so that the router will not try to send requests to it without actively probing.

Example log printed by router:
```
2024-08-08 11:37:14,390 WARNING router.py:528 -- Replica(id='q3ji1fgv', deployment='Dummy', app='default') will not be considered for future requests because it has died.
```

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/replica_scheduler/common.py [hunk 1]
```diff
@@ -241,6 +242,10 @@ def update(self, replica_id: ReplicaID, queue_len: int):
             queue_len, self._get_curr_time_s()
         )
 
+    def invalidate_key(self, replica_id: ReplicaID):
+        if replica_id in self._cache:
+            self._cache.pop(replica_id)
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/replica_scheduler/common.py:247]:
```suggestion
        self._cache.pop(replica_id, None)
```

File: python/ray/serve/tests/test_failure.py [hunk 2]
```diff
@@ -254,5 +257,65 @@ def make_blocked_request():
...
+        def check_health(self):
+            ray.get(counter.inc.remote())
+
+    h = serve.run(Dummy.options(num_replicas=2, health_check_period_s=1000).bind())
+
+    deployment_details = get_deployment_details("Dummy", _client=serve_instance)
+    replicas = [r["actor_name"] for r in deployment_details["replicas"]]
+
+    # Wait for controller to health check both replicas
+    ray.get(counter.wait.remote())
```
Comment (shrekris-anyscale) [inline python/ray/serve/tests/test_failure.py:284]: Can you add a timeout to this `ray.get` call?

**Review by shrekris-anyscale — APPROVED** [review]

File: python/ray/serve/tests/test_failure.py (hunk 2 above)
Comment (zcin) [reply python/ray/serve/tests/test_failure.py]: Added a 1s timeout

File: python/ray/serve/_private/replica_scheduler/common.py [hunk 3]
```diff
@@ -198,6 +198,7 @@ async def send_request_with_rejection(
             else:
                 return obj_ref_gen, queue_len_info
         except asyncio.CancelledError as e:
+            # HTTP client disconnected, or request was explicitly canceled
```
Comment (edoakes) [inline python/ray/serve/_private/replica_scheduler/common.py:201]:
```suggestion
            # HTTP client disconnected or request was explicitly canceled.
```

File: python/ray/serve/_private/replica_scheduler/pow_2_scheduler.py [hunk 4]
```diff
@@ -442,6 +442,15 @@ async def choose_two_replicas_with_backoff(
                     self.num_scheduling_tasks_in_backoff
                 )
 
+    def drop_replica(self, replica_id: ReplicaID):
+        self._replicas.pop(replica_id, None)
+        self._replica_id_set.discard(replica_id)
+        for id_set in self._colocated_replica_ids.values():
+            id_set.discard(replica_id)
```
Comment (edoakes) [inline python/ray/serve/_private/replica_scheduler/pow_2_scheduler.py:449]: nit: this is sort of exposing an implementation detail, would prefer to leave the details of what happens to the scheduler implementation:
```suggestion
    def on_replica_actor_died(self, replica_id: ReplicaID):
        """Describe what we do."""
        self._replicas.pop(replica_id, None)
        self._replica_id_set.discard(replica_id)
        for id_set in self._colocated_replica_ids.values():
            id_set.discard(replica_id)
```

File: python/ray/serve/_private/replica_scheduler/pow_2_scheduler.py [hunk 5]
```diff
@@ -521,12 +530,15 @@ async def _probe_queue_lens(
...
                 if isinstance(t.exception(), ActorDiedError):
-                    self._replicas.pop(replica.replica_id, None)
-                    self._replica_id_set.discard(replica.replica_id)
-                    for id_set in self._colocated_replica_ids.values():
-                        id_set.discard(replica.replica_id)
+                    self.drop_replica(replica.replica_id)
                     msg += " This replica will no longer be considered for requests."
+                # Replica is temporarily unavailable because of network issues, or
+                # replica has died but GCS is down so ActorUnavailableError will
+                # be raised until GCS recovers. For the time being, invalidate
```
Comment (edoakes) [inline python/ray/serve/_private/replica_scheduler/pow_2_scheduler.py:537]:
```suggestion
                # be raised until the GCS recovers. For the time being, invalidate
```

File: python/ray/serve/_private/router.py [hunk 6]
```diff
@@ -524,6 +525,22 @@ async def schedule_and_send_request(
...
+                    f"{replica.replica_id} will not be considered for future "
+                    "requests because it has died."
+                )
+                raise
+            except ActorUnavailableError:
+                # There are network issues, or replica has died but GCS is down so
+                # ActorUnavailableError will be raised until GCS recovers. For the
+                # time being, invalidate the cache entry so that we don't try to
+                # send requests to this replica without actively probing.
+                self._replica_scheduler.invalidate_cache_entry(replica.replica_id)
```
Comment (edoakes) [inline python/ray/serve/_private/router.py:542]: ditto to my other comment, suggest to make this `self._replica_scheduler.on_replica_actor_unavailable`

File: python/ray/serve/tests/unit/test_router.py [hunk 7]
```diff
@@ -498,6 +520,94 @@ async def test_max_queued_requests_updated(
...
+        d_id = DeploymentID(name="test")
+        r_id = ReplicaID(unique_id="r1", deployment_id=d_id)
+
+        fake_replica_scheduler.set_replica_to_return(
+            FakeReplica(r_id, error=ActorDiedError())
+        )
+        with pytest.raises(ActorDiedError):
+            await router.assign_request(dummy_request_metadata())
+
+        assert r_id in fake_replica_scheduler._dropped_replicas
```
Comment (edoakes) [inline python/ray/serve/tests/unit/test_router.py:546]: shouldn't reach into private methods/fields

**Comment by edoakes** [comment]
High level looks good, only stylistic comments

File: python/ray/serve/_private/replica_scheduler/pow_2_scheduler.py (hunk 4 above)
Comment (zcin) [reply python/ray/serve/_private/replica_scheduler/pow_2_scheduler.py:449]: Ah nice, this is a lot cleaner! renamed to `on_replica_actor_died`

File: python/ray/serve/tests/unit/test_router.py (hunk 7 above)
Comment (zcin) [reply python/ray/serve/tests/unit/test_router.py]: fixed!

**Review by edoakes — APPROVED** [review]

**Pull request merged**